source/extensions/filters/http/jwt_authn/jwks_async_fetcher.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/extensions/filters/http/jwt_authn/v3/config.pb.h"
+#include "envoy/server/factory_context.h"
+
+#include "common/common/logger.h"
+#include "common/init/target_impl.h"
+
+#include "extensions/filters/http/common/jwks_fetcher.h"
+#include "extensions/filters/http/jwt_authn/stats.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+/**
+ *  CreateJwksFetcherCb is a callback interface for creating a JwksFetcher instance.
+ */
+using CreateJwksFetcherCb = std::function<Common::JwksFetcherPtr(Upstream::ClusterManager&)>;
```

Comment:
Why is this callback needed? I traced it back up to here, all it does is create a unique_ptr from the param. Can't this `AsyncFetcher` create it instead of having the cb passed across multiple layers?

https://github.com/envoyproxy/envoy/blob/a12869fa9e9add4301a700978d5489e6a0cc0526/source/extensions/filters/http/common/jwks_fetcher.cc#L120

Reply:
It is for unit-test mocking.   You either have to define a factory class,   or a callback function so it can be mocked.

Reply:
BTW, this type seems to be defined twice in the same namespace. Move to some common place, perhaps?

Reply:
Done

Reply:
Both headers also include "common/jwks_fetcher.h" where `Common::JwksFetcherPtr` is defined. Would be logical to have the definition of `Common::JwksFetcherPtr`'s factory interface in the same header.